### PR TITLE
Added support for an own Kusama node and more

### LINKED ISSUE #1920: Add ability to use local Kusama node
## Abstract

This is part of the work required to implement the [Kusama integration](https://github.com/rotki/rotki/issues/1081) Milestone 1.

## Task

Add a new setting for the users to provide a custom endpoint if they run their own node and use it for balance query instead of the defaults.

### LINKED ISSUE #2084: SubstrateInterface can raise TypeError
## Problem Definition

As seen in this [test failure](https://github.com/rotki/rotki/pull/2030/checks?check_run_id=1697843338) it seems that the library we are using for substrate interface connections can under some circumstances raise a `TypeError`.

I think this is a bug in their implementation since from the stack trace it looks like the `response` is not checked against being `None`.

```
2021-01-13T21:16:33.0336089Z [13/01/2021 20:59:12 UTC] DEBUG rotkehlchen.chain.substrate.manager: Kusama querying last block url=https://kusama-rpc.polkadot.io/
2021-01-13T21:16:33.0337198Z Traceback (most recent call last):
2021-01-13T21:16:33.0343568Z   File "src/gevent/greenlet.py", line 818, in gevent._greenlet.Greenlet.run
2021-01-13T21:16:33.0344678Z   File "/home/runner/work/rotki/rotki/rotkehlchen/chain/substrate/manager.py", line 235, in _connect_node
2021-01-13T21:16:33.0345541Z     last_block = self._check_node_synchronization(node_interface)
2021-01-13T21:16:33.0346495Z   File "/home/runner/work/rotki/rotki/rotkehlchen/chain/substrate/manager.py", line 177, in _check_node_synchronization
2021-01-13T21:16:33.0347397Z     last_block = self._get_last_block(node_interface=node_interface)
2021-01-13T21:16:33.0348261Z   File "/home/runner/work/rotki/rotki/rotkehlchen/chain/substrate/manager.py", line 349, in _get_last_block
2021-01-13T21:16:33.0349024Z     block_hash=node_interface.get_chain_head(),
2021-01-13T21:16:33.0350332Z   File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/substrateinterface/base.py", line 732, in get_block_number
2021-01-13T21:16:33.0351403Z     return int(response['result']['number'], 16)
2021-01-13T21:16:33.0352137Z TypeError: 'NoneType' object is not subscriptable
2021-01-13T21:16:33.0354258Z 2021-01-13T20:59:12Z <Greenlet at 0x7fab325ca4d0: <bound method SubstrateManager._connect_node of <rotkehlchen.chain.substrate.manager.SubstrateManager object at 0x7fab2c701d50>>(node=<KusamaNodeName.PARITY: 0>)> failed with TypeError
```

## Task

1. Handle this `TypeError` on our side and log a proper error and not fail with unhandled exception
2. Raise this issue with the library we use (https://github.com/polkascan/py-substrate-interface) as they either have a bug, or we are using them in the wrong way.

### PR DESCRIPTION
**Pull request by vnavascues**
Closes #1920
Closes #2084

Also fixes KSM null totals when the user has not imported any KSM address.

### CODE REVIEW

File: rotkehlchen/chain/substrate/manager.py [hunk 1]
```diff
@@ -363,15 +403,27 @@ def _get_nodes_call_order(self) -> NodesCallOrder:
...
+        nodes_sorted.extend(
+            sorted(
+                cast(Iterable, node_attributes_map.items()),
+                key=lambda item: -item[1].weight_block,
+            ),
         )
+        return nodes_sorted
 
-    def _get_node_interface(self, node: NodeName) -> SubstrateInterface:
+    def _get_node_endpoint(self, node: NodeName) -> str:
```
Comment (LefterisJP) [inline rotkehlchen/chain/substrate/manager.py:423]: Where is this used?

File: rotkehlchen/chain/substrate/manager.py [hunk 2]
```diff
@@ -363,15 +403,27 @@ def _get_nodes_call_order(self) -> NodesCallOrder:
...
-        NB: currently preference is based on how close they are to the chain
-        height requested via Subscan API, the higher 'weight_block' the better.
+        Own node always has preference, then are ordered depending on how close
+        they are to the chain height; the higher 'weight_block' the better.
         """
-        return sorted(
-            cast(Iterable, self.available_node_attributes_map.items()),
-            key=lambda item: -item[1].weight_block,
+        nodes_sorted: NodesCallOrder = []
+        node_attributes_map = self.available_node_attributes_map.copy()
```
Comment (LefterisJP) [inline rotkehlchen/chain/substrate/manager.py:410]: Hmm this is not a good idea. Every time the `_get_nodes_call_order()` youare copying the map and then popping own node from it and putting it in front.

Instead what you should do is move this logic to the moment the own node connects. At that point put it to the top of the sorted list.

File: rotkehlchen/chain/substrate/manager.py [hunk 3]
```diff
@@ -541,3 +593,20 @@ def get_last_block(
         trying with all the available nodes.
         """
         return self._get_last_block(node_interface)
+
+    def set_rpc_endpoint(self, endpoint: str) -> Tuple[bool, str]:
+        """Attempt to set the RPC endpoint for the user's own node.
+        If connection at endpoint is successful it will set `own_rpc_endpoint`.
+        """
+        if endpoint == '':
+            current_own_rpc_endpoint = getattr(self, 'own_rpc_endpoint', None)
```
Comment (LefterisJP) [inline rotkehlchen/chain/substrate/manager.py:602]: Why do this? Just set `self.own_rpc_endpoint = None` in the constructor. No need for `getattr`

File: rotkehlchen/chain/substrate/manager.py [hunk 4]
```diff
@@ -541,3 +593,20 @@ def get_last_block(
...
+            log.debug(f'{self.chain} removing own node at endpoint: {current_own_rpc_endpoint}')
+            self.available_node_attributes_map.pop(SubstrateOwnNodeName.OWN, None)
+            self.own_rpc_endpoint = ''
+            return True, ''
+
+        log.debug(f'{self.chain} setting own node at endpoint: {endpoint}')
```
Comment (LefterisJP) [inline rotkehlchen/chain/substrate/manager.py:608]: move this into the `_connect_node` function

File: rotkehlchen/chain/substrate/typing.py [hunk 5]
```diff
@@ -10,6 +10,23 @@
 BlockNumber = NewType('BlockNumber', int)
 
 
+class SubstrateOwnNodeName(Enum):
```
Comment (LefterisJP) [inline rotkehlchen/chain/substrate/typing.py:13]: What is this type for? Why on earth would you ever make an Enum with 1 option? I don't get this ... what are you doing here?

This is just another node to connect to, so it should be part of the `KusamaNodeName` Enum.

**Review by LefterisJP — COMMENTED** [review]
Some comments on the code.

API docs also need to be updated for the new setting

**Comment by LefterisJP** [comment]
Also you should rebase this on develop

File: rotkehlchen/chain/substrate/typing.py (hunk 5 above)
Comment (vnavascues) [reply rotkehlchen/chain/substrate/typing.py:13]: My point of view was that "OWN" should be a singleton or a unique object, and not another option per each substrate chain we support. I've implemented it as an Enum for convenience: same methods and it behaves like the other public node options.

A specific OWN option per chain Enum (e.g. KusamaNodeName, PolkadotNodeName) requires to extend any node logic for taking into account the chain the instance is connected to (if Kusama check/use KusamaNodeName.OWN, if Polkadot check/use PolkadotNodeName.OWN). In the current implementation the instance does not need to know which chain is connected to.

I will change it though

File: rotkehlchen/chain/substrate/manager.py (hunk 3 above)
Comment (vnavascues) [reply rotkehlchen/chain/substrate/manager.py:602]: This is one of the things I addressed in the other feature I've been working. I will include it here.

File: rotkehlchen/chain/substrate/manager.py (hunk 2 above)
Comment (vnavascues) [reply rotkehlchen/chain/substrate/manager.py:410]: Agree. To be honest we don't know yet about public substrate nodes behaviour and we don't have too many public samples where to choose :D... just parity.

File: rotkehlchen/chain/substrate/typing.py (hunk 5 above)
Comment (LefterisJP) [reply rotkehlchen/chain/substrate/typing.py:13]: It wouldn't be KusamaNodeName, PolkadotNodeName. Just `NodeName` or `SubstrateNodeName`. This extra class is just not required.

File: rotkehlchen/chain/substrate/manager.py (hunk 3 above)
Comment (LefterisJP) [reply rotkehlchen/chain/substrate/manager.py:602]: yeah include it and when we merge this, your other work can be rebased on this

File: docs/changelog.rst [hunk 6]
```diff
@@ -2,7 +2,8 @@
 Changelog
 =========
 
-* :feature:`1662` Users are now able to manually input ledger actions such as Income, Donation, Loss, Expense, Dividends Income
+* :feature:`1920` Rotki now supports your own Kusama node along the public ones.
```
Comment (LefterisJP) [inline docs/changelog.rst:5]: Can you change the wording here in a subsequent PR? "Rotki now supports addition of a custom Kusama endpoint"

File: rotkehlchen/chain/substrate/typing.py [hunk 7]
```diff
@@ -13,14 +13,22 @@
 class KusamaNodeName(Enum):
     """Public nodes for Kusama.
     """
-    PARITY = 0
+    OWN = 0
+    PARITY = 1
```
Comment (LefterisJP) [inline rotkehlchen/chain/substrate/typing.py:17]: What happened to the web3 node?

File: rotkehlchen/chain/substrate/typing.py [hunk 8]
```diff
@@ -13,14 +13,22 @@
...
 
     def __str__(self) -> str:
+        if self == KusamaNodeName.OWN:
+            return 'own node'
         if self == KusamaNodeName.PARITY:
             return 'parity'
         raise AssertionError(f'Unexpected KusamaNodeName: {self}')
 
     def endpoint(self) -> str:
+        if self == KusamaNodeName.OWN:
```
Comment (LefterisJP) [inline rotkehlchen/chain/substrate/typing.py:27]: This is not bad, so no need to change it.

An alternative would be to pass the rpc endpoint as an optional argument here, so if it's passed (which you would do for own node) then you use the given url. This is what we do with ethereum.

**Review by LefterisJP — APPROVED** [review]
LGTM. Left some comments to keep in mind.

File: docs/changelog.rst (hunk 6 above)
Comment (vnavascues) [reply docs/changelog.rst:5]: yep!

File: rotkehlchen/chain/substrate/typing.py (hunk 7 above)
Comment (vnavascues) [reply rotkehlchen/chain/substrate/typing.py:17]: Only works under web sockets and from time to time... sometimes SSL problems, sometimes connection closed. Parity is the most reliable one using the `SubstrateInterface`.

File: rotkehlchen/chain/substrate/typing.py (hunk 8 above)
Comment (vnavascues) [reply rotkehlchen/chain/substrate/typing.py:27]: I saw it but also realised that `endpoint` in NodeName is not optional but required. I thought that if I have it in the instance why passing it around. Anyway, these are just nuances :)

**Pull request merged**